src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -44,69 +50,105 @@ class IcebergParquetValueParser {
-      PrimitiveType primitiveType = type.asPrimitiveType();
-      switch (primitiveType.getPrimitiveTypeName()) {
-        case BOOLEAN:
-          int intValue =
-              DataValidationUtil.validateAndParseBoolean(
-                  type.getName(), value, insertRowsCurrIndex);
-          value = intValue > 0;
-          stats.addIntValue(BigInteger.valueOf(intValue));
-          estimatedParquetSize += ParquetBufferValue.BIT_ENCODING_BYTE_LEN;
-          break;
-        case INT32:
-          int intVal = getInt32Value(value, primitiveType, insertRowsCurrIndex);
-          value = intVal;
-          stats.addIntValue(BigInteger.valueOf(intVal));
-          estimatedParquetSize += 4;
-          break;
-        case INT64:
-          long longVal = getInt64Value(value, primitiveType, defaultTimezone, insertRowsCurrIndex);
-          value = longVal;
-          stats.addIntValue(BigInteger.valueOf(longVal));
-          estimatedParquetSize += 8;
-          break;
-        case FLOAT:
-          float floatVal =
-              (float)
-                  DataValidationUtil.validateAndParseReal(
-                      type.getName(), value, insertRowsCurrIndex);
-          value = floatVal;
-          stats.addRealValue((double) floatVal);
-          estimatedParquetSize += 4;
-          break;
-        case DOUBLE:
-          double doubleVal =
-              DataValidationUtil.validateAndParseReal(type.getName(), value, insertRowsCurrIndex);
-          value = doubleVal;
-          stats.addRealValue(doubleVal);
-          estimatedParquetSize += 8;
-          break;
-        case BINARY:
-          byte[] byteVal = getBinaryValue(value, primitiveType, stats, insertRowsCurrIndex);
-          value = byteVal;
-          estimatedParquetSize +=
-              ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN + byteVal.length;
-          break;
-        case FIXED_LEN_BYTE_ARRAY:
-          byte[] fixedLenByteArrayVal =
-              getFixedLenByteArrayValue(value, primitiveType, stats, insertRowsCurrIndex);
-          value = fixedLenByteArrayVal;
-          estimatedParquetSize +=
-              ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN + fixedLenByteArrayVal.length;
-          break;
-        default:
+      if (type.isPrimitive()) {
+        if (!statsMap.containsKey(path)) {
           throw new SFException(
-              ErrorCode.UNKNOWN_DATA_TYPE,
-              type.getLogicalTypeAnnotation(),
-              primitiveType.getPrimitiveTypeName());
+              ErrorCode.INTERNAL_ERROR,
+              String.format("Stats not found for column: %s", type.getName()));
+        }
+        RowBufferStats stats = statsMap.get(path);
+        estimatedParquetSize += ParquetBufferValue.DEFINITION_LEVEL_ENCODING_BYTE_LEN;
+        estimatedParquetSize +=
+            isdDescendantsOfRepeatingGroup
+                ? ParquetBufferValue.REPETITION_LEVEL_ENCODING_BYTE_LEN
```

Comment:
is this only needed for primitive typed values? What if its an array of objects, do we not need this constant to be added to estimatedParquetSize for each object of the array?

Reply:
I used [this paper](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/36632.pdf) as reference to understand repetition level and definition level encoding. And from [this article](https://blog.x.com/engineering/en_us/a/2013/dremel-made-simple-with-parquet), only descendants of a repeating group needs to record repetition level. We only need these on leaf values.